tests/test_user_flows.py

```diff
@@ -1,36 +1,202 @@
+from bs4 import BeautifulSoup
+from flask import session, url_for
 import pytest
 
+from app.main.views.auth import idp_profiles
+
+
+email_idp_name = [
+    ('test@digital.cabinet-office.gov.uk', 'gds-google',
+     'Government Digital Service'),
+    ('test.test@cabinetoffice.gov.uk', 'co-digital', 'Cabinet Office'),
+    ('test@sso.civilservice.digital', 'ad-saml', 'Civil Service Digital'),
+]
+
 
 class TestUserFlows(object):
 
-    @pytest.mark.xfail
-    def test_confirm_email_address(self):
-        assert False
+    def assertMetaRefresh(self, response, sec, url):
+        content = BeautifulSoup(response.data.decode('utf-8'), 'html.parser')
+        meta = content.find(
+            'meta',
+            attrs={
+                'http-equiv': 'refresh',
+                'content': '%s; url=%s' % (sec, url)})
+        assert meta is not None
+
+    @pytest.mark.parametrize("email, idp, name", email_idp_name)
+    def test_confirm_email_address(self, app_, email, idp, name):
+        url = url_for('main.request_email_address')
+        data = {
+            'email_known': 'yes',
+            'email_address': email
+        }
+
+        with app_.test_client() as c:
+            assert c.get(url).status_code == 200
+
+            resp = c.post(url, data=data)
+            assert resp.status_code == 302
+            assert resp.location.endswith(url_for('main.confirm_dept'))
+            assert session['email_address'] == email
+            assert session['suggested_idp'] == idp
+            assert session['department_name'] == name
```

Comment:
These assertions about the session are too tightly coupled to the implementation. These functional tests should probably just assert that the client see the expected response.